Keyboard Shortcuts for TM:PE

kvakvs opened the pull request:
# Note to reviewers
Please review with "Ignore whitespace" (⚙️ button above the diff view)

## Changes

GUI settings page for keyboard shortcuts.
TMPE main window and 6 most used tools can be used with keyboard.
Settings are automatically saved by ColossalFramework.
Tooltips are shown on the menu buttons for the current keyboard shortcut.
FIXME: Tooltip on main crown button is not shown.
Stay in lane Shift+S shortcut does work

## Checklist

- [x] Secondary keyboard shortcut
- [ ] ~+click mouse combinations allowing to select only modifier keys~ Moved to #402 
- [x] Tooltips without a key should say nothing
- [x] 👑 button tooltip is missing

![image](https://user-images.githubusercontent.com/288724/59557736-be062c80-8fe1-11e9-8832-074c381f4799.png)
![image](https://user-images.githubusercontent.com/288724/59557742-d6764700-8fe1-11e9-9f97-70d3f0e7a04c.png)

kvakvs commented:
Working to add another commit to this, which will also allow customizing Shift+S stay in lane key shortcut. At the time of writing, it works if it is not Shift+S for me, so something else is blocking it.

krzychu124 commented:
I think you forgot to add `KeyMapping.cs` to csproj because our CI is failing 😉

FireController1847 commented on TLM/TLM/Resources/lang.txt:
```diff
@@ -231,4 +231,12 @@ Also_apply_to_left/right_turns_between_one-way_streets Also apply to left & righ
+Keyboard_use_priority_signs_tool Use 'Priority Signs' Tool
+Keyboard_use_junction_restrictions_tool Use 'Junction Restrictions' Tool
+Keyboard_use_speed_limits_tool Use 'Speed Limits' Tool
+Keyboard_lane_connector_stay_in_lane Lane connector: Stay in lane
```
"Keybind" would probably work better here than "Keyboard"

kvakvs replied:
Changed

FireController1847 commented on TLM/TLM/Resources/lang.txt:
```diff
@@ -59,7 +59,7 @@ All_vehicles_may_ignore_lane_arrows All vehicles may ignore lane arrows
 Reckless_driving Reckless driving
 TMPE_Title Traffic Manager: President Edition
-Settings_are_defined_for_each_savegame_separately Settings are defined for each savegame separately
+Settings_are_defined_for_each_savegame_separately Settings are defined for each savegame separately. Please open settings while in-game.
```
Is there a point of adding this line? I feel as if this is strongly implied and people can figure that out.

aubergine10 replied:
I think it's needed for disambiguation, especially has "read only mod options" (when accessed from main menu) is such a weird concept considering how all other mods work. The original text made it sound like global settings would still have an effect but the save game takes precedence, which is not what happens.

IMO, better text might be something like:

> Settings can only be changed whilst in-game, and are unique to each save game.

kvakvs replied:
Changed to the suggested.

FireController1847 replied:
The changed version works much better.

FireController1847 commented on TLM/TLM/State/Options.cs:
```diff
@@ -387,6 +387,24 @@ public class Options : MonoBehaviour {
+                        // KEYBOARD
+                        ++tabIndex;
+
+                        AddOptionTab(tabStrip, Translation.GetString("Keyboard"));
```
Maybe I'm missing something, but the "Keyboard" translation was not added to the list. Also, same suggestion here, rename to "Keybinds."

kvakvs replied:
Done

krzychu124 commented on TLM/TLM/Resources/lang_pl.txt:
```diff
@@ -231,4 +231,13 @@ Also_apply_to_left/right_turns_between_one-way_streets Uwzględnia również skr
+Keybind_use_junction_restrictions_tool Use 'Junction Restrictions' Tool
+Keybind_use_speed_limits_tool Use 'Speed Limits' Tool
+Keybind_lane_connector_stay_in_lane Lane connector: Stay in lane
+Keybinds Keybinds
```
You can replace those translation keys 😉 

Keybind_toggle_TMPE_main_menu `Menu TM:PE`
Keybind_toggle_traffic_lights_tool `Narzędzie 'Przełącznik Sygnalizacji świetlnej' `
Keybind_use_lane_arrow_tool `Narzędzie 'Strzałki pasów ruchu'`
Keybind_use_lane_connections_tool `Narzędzie 'Połącz pasy ruchu'`
Keybind_use_priority_signs_tool `Narzędzie 'Znaki pierwszeństwa'`
Keybind_use_junction_restrictions_tool `Narzędzie 'Ograniczenia na skrzyżowaniu'`
Keybind_use_speed_limits_tool `Narzędzie 'Limity prędkości'`
Keybind_lane_connector_stay_in_lane `Połącz pasy ruchu: Pozostań na pasie`
Keybinds `Skróty klawiszowe`

kvakvs replied:
Added

krzychu124 replied:
https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/blob/d68efff147db6802173f08e1c4afb8d4195b7a36/TLM/TLM/Resources/lang_pl.txt#L66

Please remove ` (dodatkowe informacje) ` from that translation key to be able to see keybinds tab while Polish lang selected 😉

aubergine10 replied:
Just tested Polish locale and seems this change was already made.

kvakvs commented:
I need some help with Shift+S shortcut, it seemed to not work for me.
It works if I use something other than Shift+S though.

aubergine10 commented:
I'm not sure how the game's own keymapping system works. Maybe it treats every shortcut as globally unique, rather than tool specific?

krzychu124 commented:
I will try to check what is going there when I'll be back home :wink:

krzychu124 commented:
Ok, I think I found possible cause... https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/blob/1f671ed2d0025b7f3a0be551d4e8c5d5c0dcf4bc/TLM/TLM/UI/SubTools/LaneConnectorTool.cs#L191
This Event.current is not the same even as that triggered in OnGui() method... Inside OnGui() Event has KeyCode.S but this `Event.Current` doesn't (it has character field set to 83'S', instead of KeyCode enum).

Main part of source code of `public bool IsPressed(Event e)`
```cs
// keyCode is our KeyCode to test against
// e.keyCode comes from Event
return keyCode != KeyCode.None && e.keyCode == keyCode && ... //and more checks there
```

Another issue, when I switched to Polish lang, tabs are wider and Keybinds tab is not accessible/not visible at all. (I will try to change translation of Maintenance tab in Polish)

Additionally sometimes some key combination Ctrl+Alt+C is locking in-game movement with WASD and mouse middle button - I can't rotate camera. Maybe not related with this PR

krzychu124 commented on TLM/TLM/UI/MainMenu/MainMenuPanel.cs:
```diff
@@ -206,5 +207,37 @@ public class SizeProfile {
+                        // Check the key and then click the corresponding button
+                        if (!UIView.HasModalInput()
+                            && !UIView.HasInputFocus()) {
+                                if (OptionsKeymapping.KeyToggleTrafficLightTool.IsPressed(Event.current)) {
```
I think that checking if key is pressed while TM:PE menu isn't currently open seems to be pointless.

kvakvs replied:
I believe that OnGUI will not run for an element if it is not visible.
I do not understand what should be changed here.

krzychu124 replied:
I've just checked what is going on when we hiding panel and:
Unity docs: `If the MonoBehaviour's enabled property is set to false, OnGUI() will not be called`
but...
In this case `OnGUI` is called, because our `MainMenuPanel: UIPanel` is not `disabled` but `hidden` using internal flag `UIComponent.isVisible`

I suggest adding `this.isVisible` just before `!UIView.HasModalInput()` check and I think we can merge because everything is working right.

@aubergine10 what do you think?

kvakvs commented:
Fixed Shift+S shortcut now works, also done Delete/Bkspace in the same way, but did not add a new saved inputkey for it.

aubergine10 commented:
Just pushed a commit that renames "Policies & Restrictions" tab to just "Policies". This should solve the issue with not enough space on the tab bar.

I haven't changed the locale key, that can wait until we do the transition to xml locale files. See commit description for more infos.

aubergine10 commented:
@kvakvs Can you check through the list of changed files on this PR and see if any of them are also altered by PR #260 (Migration to Harmony) - if so this PR #382 will need to wait until _after_ the harmony update is done IMO.

kvakvs commented:
@aubergine10 Harmony PR changes 237 files and 10k lines, but i don't see a significant overlap.
There is overlap of course but in a few minor locations.

aubergine10 commented:
Which minor locations specifically? We need to assess what the impact will be of merging this PR 382 prior to Harmony. If it's not going to cause much pain then I think this PR 382 could potentially get rolled in to the v10.21 milestone (next planned release).

aubergine10 commented:
Is there a way to remove shortcuts I don't use? For example, if I rarely use timed traffic lights, I don't want keyboard shortcut lingering for that, so I'd want some way to clear it. Maybe a small ( x ) button at side of the current shortcut box to clear a shortcut?

Also, I think we should be defaulting to not defining the shortcuts - there are lots of other existing mods with customisable shortcuts so pushing a TM:PE update with pre-defined shortcut keys could conflict and cause problems for users. It might be better to leave the shortcut boxes empty by default so users can define their own _and will thus know the reason for any shortcut conflicts_.

For the TM:PE menu bar, however, I think we could potentially pre-define its shortcut. When I thought about shortcuts a few months back, I was thinking of maybe using `Shift + ;` (in other words: `:` like the colon in `TM:PE`) - I haven't seen anything else using that so it should be conflict-free. Of course, users can easily change it to something else if they want.

FireController1847 commented:
To remove it, instead of it being a button, do what most games do and when you click on it to change it allow the user to press "ESCAPE" to remove the shortcut

aubergine10 commented:
Yup, that would work (especially as we don't want `Esc` to ever be used as it's the hard-coded "exit from (sub)tool" button). But then user would have to know about pressing `Esc` to clear shortcut, so I still think a ( x ) button would also be useful as a visual cue (it's tool tip could mention the "Esc to clear a shortcut" maybe?)

One other thing I just realised (while working on docs for this feature)... I assume at a later date we will extend that Keyboard tab in mod options to include all the sub-tool shortcuts as well? If so, they would likely get their own sections what would become a scrolling panel. That being the case, I think we should rename the "Keyboard" _section heading_ to "Toolbar" or "Tools". And then at later date tool specific shortcuts would get their own sections like "Lane Connector Tool" or whatever.

aubergine10 commented:
And... This tab will become the go-to reference for users in terms of shortcuts. Is it possible to add a read-only shortcut to the top of the current list:

> Exit from active tool or sub-tool  `[ Esc         ]`

kvakvs commented:
![image](https://user-images.githubusercontent.com/288724/59867363-4db72c80-938e-11e9-8eca-94180ba02d24.png)

kvakvs commented:
@aubergine10 I've tried merging this onto master and then merging Harmony PR onto of it, besides of the fact that I could not compile the end result, i did not find any overlaps or conflicts in the code which I have edited. There were Harmony branch vs master branch conflicts though.

krzychu124 commented:
@kvakvs don't forget to sync RedirectionFramework submodule, and switch to customization/TMPE branch there before building mod

kvakvs commented on TLM/TLM/State/KeyMapping.cs:
```diff
@@ -0,0 +1,333 @@
+			}
+		}
+
+		protected void OnBindingMouseDown(UIComponent comp, UIMouseEventParameter p) {
```
I am not sure if we can do anything about this. There is no list of all shortcuts used by other mods to check. Maybe we can get to the shortcuts used by the game itself.

kvakvs commented on TLM/TLM/Resources/lang_pl.txt:
```diff
@@ -231,4 +231,14 @@ Also_apply_to_left/right_turns_between_one-way_streets Uwzględnia również skr
+Keybind_use_speed_limits_tool Narzędzie 'Limity prędkości'
+Keybind_lane_connector_stay_in_lane Połącz pasy ruchu: Pozostań na pasie
+Keybinds Skróty klawiszowe
+Keybind_Exit_subtool Exit tool and close TM:PE Menu
```
Updated

kvakvs commented on TLM/TLM/State/KeyMapping.cs:
```diff
@@ -0,0 +1,333 @@
+		protected void TryCreateConfig() {
+			try {
+				// Creating setting file
+				if (GameSettings.FindSettingsFileByName(KEYBOARD_SHORTCUTS_FILENAME) == null) {
```
I think we are trying to use `SavedInputKeys` before the config file was read or created. It is only created when Options window creates tab for keyboard bindings. This is a bug.

krzychu124 replied:
Nope, I checked that and if you call `GameSettings.FindSettingsFileByName` and `SettingsFile` is missing in internal dictionary of files it throws that line to log.

kvakvs commented:
![image](https://user-images.githubusercontent.com/288724/59969057-4f772080-9545-11e9-8156-6d040063bcc6.png)

TODO: We probably want to categorize our keys too, and only find conflicts within same category. Example: Lane connector uses Shift+S, i want to only check for conflicts within other Lane Connector keys and other tools are welcome to share the key.

aubergine10 commented:
Can you add a blank line between the text and the name of the other shortcut to aid readability?

kvakvs commented:
Added categories to the keys.
Now if the key belonged to anything other than "Global", like Lane Connector key, it will only search for conflicts within the Global and its own category, which will allow us to reuse same key in multiple tools.

![image](https://user-images.githubusercontent.com/288724/59969506-bc42e880-954e-11e9-93dd-880a2d5207a4.png)

kvakvs commented:
Solved the merge conflicts by joining all additions to translations together.

aubergine10 commented:
Finally got round to doing some testing on this PR...

The main TMPE button doesn't have a tooltip (thus doesn't display keybind):

![image](https://user-images.githubusercontent.com/1386719/60323272-869c6600-9979-11e9-89f7-45f9d0a88c03.png)

Stating 'None' when there is no keybind looks weird:

![image](https://user-images.githubusercontent.com/1386719/60323338-b6e40480-9979-11e9-8bdd-0fe39983a306.png)

If no keybind it would be better to just not output the keybind line of the tooltip.

Looks like a locale string is missing for this keybind:

![image](https://user-images.githubusercontent.com/1386719/60323437-0296ae00-997a-11e9-921d-5ccd8fed0d07.png)

Also, note that most Mac keyboards don't have a `Delete` key. As such, we should always treat `Delete` and `Backspace` as synonymous.

I noticed that the only way to delete an existing keyboard shortcut is to go in to 'define mode' and then press `Delete` (pressing `Esc` merely cancels 'define mode'). I still think having a little `x` button next to defined keybinds would be a better way to remove them.

Can we treat `;` and `:` as synonymous? And have `:` shown as the default shortcut for the TM:PE main menu?

Upon adding a shortcut for a tool (`S` for speeds) I expected that just pressing `S` would enter that tool, even if main menu wasn't open, but the main menu needs opening first before it can be used (I agree with this approach btw). So I was wondering if some sort of indenting would help clarify what's going on:

```
// note: I'm using [whatever] to denote read-only fields

* = Can be repeated to cycle through options

Toggle TM:PE menu bar                   :
  Toggle Traffic Lights
    Toggle lights                    [Click]
  Lane Arrows
    (wait for PR #391)
  Lane Connections
    Select node/lane                 [Click]
      Stay in lane *                Shift + S
      Reset to default               Delete
      Exit node/lane              [Right Click]
  Priority Signs
    Select node                      [Click]
      Change sign *                  [Click]
      Apply to route              Shift [+ Click]
      Exit node                   [Right Click]
  Timed Traffic Lights
    Select node                      [Click]
  Junction Restrictions
    Select node                      [Click]
      Toggle restriction *           [Click]
      Exit node                   [Right Click]
  Speed Limits
    Select / apply speed             [Click]
    Apply between junctions       Shift [+ Click]
  Vehicle Restrictions
    Toggle restriction *             [Click]
    Apply between junctions       Shift [+ Click]
  Parking Restrictions
    Toggle restriction *             [Click]
    Apply between junctions       Shift [+ Click]
  Exit tool or menu bar               [Esc]
```

When a key modifies a click, we should append ` + Click` to the keybind to make it obvious what the user has to do.

Note that the `[Right Click]` could be `[Left Click]` depending on OS mouse button settings. The game somehow automagically handles this by silently swapping left/right clicks so there must be a way to detect how the mouse is configured.

The indentation not only acts as an interaction guide for users, but it also indicates keybind scoping for developers.

kvakvs commented:
The UI uses a template object from the game, with a label and a button for keyboard shortcut. Modifying it will require recreating each row instead of instantiating that template.
Adding a second alternate shortcut might be a good idea, then we can bind Delete and Backspace.
Changing how just keyboard bindings are converted to string is not a good idea, not worth it in general without reworking the whole thing with table of keys and localization and everything.

aubergine10 commented:
> Adding a second alternate shortcut might be a good idea, then we can bind Delete and Backspace.

Yes, this would be good.

> The UI uses a template object from the game, with a label and a button for keyboard shortcut. Modifying it will require recreating each row instead of instantiating that template.

If we modify layout so that there are three columns (label, main key, alt key) could we use the alt key column for static '+ Click' on those click modifiers?

> Changing how just keyboard bindings are converted to string is not a good idea, not worth it in general without reworking the whole thing with table of keys and localization and everything.

Is this relating to keybind table or the tooltips?

kvakvs commented:
> > The UI uses a template object from the game, with a label and a button for keyboard shortcut. Modifying it will require recreating each row instead of instantiating that template.

> If we modify layout so that there are three columns (label, main key, alt key) could we use the alt key column for static '+ Click' on those click modifiers?

I think so? I need to rethink this today after work. Modifier editor will be something different from a keyboard combination editor.

> > Changing how just keyboard bindings are converted to string is not a good idea, not worth it in general without reworking the whole thing with table of keys and localization and everything.

> Is this relating to keybind table or the tooltips?

Both actually. It is that to localized string call which displays the tooltip, i do not want to go and replace single keys in that algorithm because it is large with large localized lookup tables for the whole keyboard.

aubergine10 commented:
For the tooltip, is it possible to alternate choice of what type of tooltip is used depending on presence of keybind?

Specifically, having "None" shown after a tool name, for example `Junction Restrictions \n None` makes it look like "there's no junction restrictions".

kvakvs commented:
> For the tooltip, is it possible to alternate choice of what type of tooltip is used depending on presence of keybind?
> 
> Specifically, having "None" shown after a tool name, for example `Junction Restrictions \n None` makes it look like "there's no junction restrictions".

Of course it is possible. I'd like some way of saying that there could be a tooltip but you didn't set it, instead of it just disappearing. Disappearing is probably fine too.

kvakvs commented:
I have a feeling that the options GUI only allows adding one object per row. So there might be some fancy programming required to make a Unity GUI template with the elements we want, to have it laid out horizontally.

aubergine10 commented:
For scrolling and layout in general, see IncompatibleModsPanel.cs in PR 333 (it could probably be tidier, but it should give general gist):

https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/pull/333/files#diff-9a8fe4814d86ff2f1a6af2df21667314

I assume for settings panel to scroll we'd have to just work out the width/height or something. The [Ambient Sounds Tuner (git repo)](https://github.com/bloodypenguin/csl-ambient-sounds-tuner/tree/master/CSL%20Ambient%20Sounds%20Tuner) mod has a scrolling settings panel.

krzychu124 commented:
> I have a feeling that the options GUI only allows adding one object per row

I am not sure what do you mean, but I've just checked sources and for example `ContentPanel/Savegames` uses scrollpanel and list of items (each item has a lot of different buttons and labels). 

In ModTools each entry is visible as SaveEntryTemplate(clone) which suggest that they supply tab template with saves list and then they clone entryTemplate internally for each entry. Each ContentPanel/Options tab layout seems to be defined by template of particular `Package.AssetType`. 

All option templates were probably designed and built in Unity and game only loads those templates and fill them with data. 

We have to do that manually so I think there is no limit what we could create there.

kvakvs commented:
Untabified files affected by this PR, looking better in general

- [x] Translations: added missing one, added missing lines to the template
- [x] Having no key will not show "None" in tooltips
- [ ] 👑 button tooltip is missing
- [ ] Change keybinds GUI
  - [ ] Add delete button to the keybinds GUI like ❎
  - [ ] Dual keybinds for some shortcuts
  - [ ] Add "+ Click" support for some shortcuts (allow to only select control keys)

kvakvs commented:
Got something brewing here
![image](https://user-images.githubusercontent.com/288724/60396602-f3fbf280-9b43-11e9-82b7-2ba8a6d01642.png)

aubergine10 commented:
Will test later tonight.

For the ones with two keys, is it possible to put them side by side?

```
Clear lane connections          Delete    Backspace
```

kvakvs commented:
@aubergine10 
![image](https://user-images.githubusercontent.com/288724/60401330-ec5a3f00-9b7f-11e9-974c-64880628d7b4.png)

kvakvs commented:
Current problem: Scroll panel is not really scrolling or showing any scroll elements.
Also i don't have enough scroll elements for it, so can fix that later.

aubergine10 commented:
We can do the extra keybinds stuff in a later PR to give some more time to get the scrolling working properly and wire the other shortcuts in etc.

Will be testing in about 1 hour or so.

aubergine10 commented:
I'm currently resolving the merge conflicts from recent PR merges....

aubergine10 commented:
Found an issue, just trying to replicate...

aubergine10 commented:
It's possible to close the content manager while mod options screen is still open, resulting in un-closeable mod options screen.

Steps to reproduce:

1. Main Menu > Content Manager > Mods > TM:PE > Options
2. Select keybinds tab
3. Click a keybind button (I chose one captioned `None`, specifically the `Toggle Traffic Lights` keybind, but I'm not sure if that matters)
4. Button shows `Press any key`
5. Press `Esc` - this is where stuff starts getting weird:
    * Button seems de-selected, but continues to show `Press any key`
6. It seems that at this point the mod options modal ceases trapping keyboard or mouse input
7. I can click through it on right side to select mods in the content manager behind it
8. I can close the content manger, but the mod options dialog stays open

It's a bit hit and miss. Also, I just managed to get same sort of issue by assigning a keypress - I tried a few keys first that were already assigned to stuff; the "already assigned" warning appeared (correct) and the button went back to `None` after closing the warning (correct). However, I then choose a key that wasn't assigned to anything - the button remained `Press any key` and the entire options dialog became non-responsive (no longer trapping keyboard or mouse).

[output_log.txt](https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/files/3346819/output_log.txt)

aubergine10 commented:
Minor issue: Superfluous `X` text label over the "remove keybind" buttons.

kvakvs commented:
I think i've managed to fix the button stuck in press any key, by adding a callback on modal pop (the edit keybind button temporarily becomes a modal, and now when it closes, the edit mode will be ended).

krzychu124 commented:
I've got a new bug. The same problem as @aubergine10 but other trigger. 
I've set the same key shortcut `C` or `N` (not used by game itself) for more than one feature. 
No warning about keybind duplicate. Couldn't close options dialog

```cs
NullReferenceException: Object reference not set to an instance of an object
  at TrafficManager.State.Keybinds.KeybindSetting.NotifyKeyChanged ()
```
and then:
```cs
InvalidOperationException: Trying to close a modal panel not on the top of the modal stack
at ColossalFramework.UI.UIDynamicPanels.Hide()
```

aubergine10 commented:
I've just noticed with the mod incompatibility checker that if you `Esc` out of it, back to main menu, you can't `Esc` to exit the game. So maybe something changed in the modal popup system recently?

kvakvs commented:
@aubergine10 so far, it is due to terrible modal stack handling by me. I am fixing it.
Does incompatible mods dialog set modal and not clear it at the right time? Might be a similar problem.

aubergine10 commented:
Looking at the code, the mod checker _should_ be clearing the modal properly, and it certainly works from a visual perspective most of the time, however @krzychu124 is going to be digging in to that tomorrow to make sure. He discovered today that sometimes the modal is blurred, like it has the wrong z-order and appears behind the "blur" effect thing.

aubergine10 commented:
Just reviewing changed files and noticed the Russian translations were updated - was that you Kvakvs? (for changelog mention)

aubergine10 commented on TLM/TLM/State/Keybinds/Keybind.cs:
```diff
@@ -0,0 +1,79 @@
+        public static bool IsEmpty(InputKey sample) {
+            var noKey = SavedInputKey.Encode(KeyCode.None, false, false, false);
+            return sample == SavedInputKey.Empty
+                   || sample == noKey;
```
Readability: Put on same line as `return` statement?

kvakvs replied:
Joined lines into one

krzychu124 commented on TLM/TLM/Resources/lang_pl.txt:
```diff
@@ -233,10 +233,24 @@ Ignore_disabled_mods Ignoruj wyłączone mody
+Keybind_lane_connector_delete Oczyść połączenia pasów ruchu
+Keybinds Skróty klawiszowe
+Keybind_Exit_subtool Wyłącz narzędzie i zamknij TM:PE menu
+Keybind_conflict The key you've selected is conflicting with another shortcut key. Please choose something else.
```
```suggestion
Keybind_conflict Wybrany skrót klawiszowy jest już używany. Wybierz inną kombinację.
```

kvakvs replied:
Added

kvakvs replied:
> I noticed in code some stuff about row striping, but didn't see any on the UI? (not important, just mentioning it in case the rows were supposed to be striped)

The row striping code is still there. But in the old dialog the default background was grey-ish, and row striping code just cleared texture on some lines. In the new dialog i've no idea what the background sprite name is, so they are all black (i.e. not striped). Have to go into ModTool or UIHelper source and find the background sprite name which is used in other dialogs.

krzychu124 commented on TLM/TLM/Resources/lang_pl.txt:
```diff
@@ -233,10 +233,24 @@ Ignore_disabled_mods Ignoruj wyłączone mody
+Keybinds Skróty klawiszowe
+Keybind_Exit_subtool Wyłącz narzędzie i zamknij TM:PE menu
+Keybind_conflict The key you've selected is conflicting with another shortcut key. Please choose something else.
+Keybind_category_Global Global keys
```
```suggestion
Keybind_category_Global Główne
```

kvakvs replied:
Copied by hand

krzychu124 commented on TLM/TLM/Resources/lang_pl.txt:
```diff
@@ -233,10 +233,24 @@ Ignore_disabled_mods Ignoruj wyłączone mody
+Keybind_Exit_subtool Wyłącz narzędzie i zamknij TM:PE menu
+Keybind_conflict The key you've selected is conflicting with another shortcut key. Please choose something else.
+Keybind_category_Global Global keys
+Keybind_category_LaneConnector Lane Connector Tool keys
```
```suggestion
Keybind_category_LaneConnector Narzędzie: Połącz pasy ruchu
```

kvakvs replied:
Copied by hand

krzychu124 commented on TLM/TLM/Resources/lang_pl.txt:
```diff
@@ -233,10 +233,24 @@ Ignore_disabled_mods Ignoruj wyłączone mody
+Keybind_use_junction_restrictions_tool Narzędzie 'Ograniczenia na skrzyżowaniu'
+Keybind_use_speed_limits_tool Narzędzie 'Limity prędkości'
+Keybind_lane_connector_stay_in_lane Połącz pasy ruchu: Pozostań na pasie
+Keybind_lane_connector_delete Oczyść połączenia pasów ruchu
```
```suggestion
Keybind_lane_connector_delete Usuń połączenia pasów ruchu
```

kvakvs replied:
Copied by hand

aubergine10 commented on TLM/TLM/State/Keybinds/Keybind.cs:
```diff
@@ -0,0 +1,79 @@
+        /// </summary>
+        /// <param name="k">The key</param>
+        /// <returns>The shortcut, example: "Ctrl + Alt + H"</returns>
+        public static string Str(SavedInputKey k) {
```
Could use a better method name

kvakvs replied:
Renamed to `ToLocalizedString`

aubergine10 commented on TLM/TLM/State/Keybinds/KeybindSetting.cs:
```diff
@@ -0,0 +1,114 @@
+        /// </summary>
+        /// <param name="prefix">Prefix will be added if any key is not empty</param>
+        /// <returns>String tooltip with the key shortcut or two</returns>
+        public string Str(string prefix = "") {
```
More descriptive method name?

kvakvs replied:
Renamed to `ToLocalizedString` just as the one in `Keybind.cs`

aubergine10 commented:
One of the German locale strings is too long:

![image](https://user-images.githubusercontent.com/1386719/60532571-7b3c9800-9cf5-11e9-8aeb-7cd03676383c.png)

aubergine10 commented on TLM/TLM/State/Options.cs:
```diff
@@ -1,516 +1,466 @@
+            removeParkedVehiclesBtn = maintenanceGroup.AddButton(Translation.GetString("Remove_parked_vehicles"), onClickRemoveParkedVehicles) as UIButton;
 #if DEBUG
-			allRelaxedToggle = atJunctionsGroup.AddCheckbox(Translation.GetString("All_vehicles_may_ignore_lane_arrows"), allRelaxed, onAllRelaxedChanged) as UICheckBox;
+            resetSpeedLimitsBtn = maintenanceGroup.AddButton(Translation.GetString("Reset_custom_speed_limits"), onClickResetSpeedLimits) as UIButton;
```
Is this debug feature something added by this PR or the earlier PR that added mph speeds?

kvakvs replied:
The reformatting and regrouping the code shifted lines, making diff really confusing.
I paid greater attention to not modify the code in here, just untabify and split into smaller functions.
So this was existing.

aubergine10 commented:
Keybind options tab UI feels very solid. Tried everything I can think of and so far not been able to break it. It feels nice and responsive too, and looks really great.

I noticed in code some stuff about row striping, but didn't see any on the UI? (not important, just mentioning it in case the rows were supposed to be striped)

kvakvs commented on TLM/TLM/State/Keybinds/KeybindSetting.cs:
```diff
@@ -0,0 +1,114 @@
+using JetBrains.Annotations;
```
It is published on NuGet, and contains informative annotations such as `[NotNull]` which contain no code but rather affect compiler warnings. Might remove it too, but i liked the idea.

aubergine10 replied:
Just tested the build and didn't cause any issues, so I'm fine with it if the other devs are.

kvakvs commented on TLM/TLM/State/Keybinds/KeybindSetting.cs:
```diff
@@ -0,0 +1,114 @@
+using UnityEngine;
+
+namespace TrafficManager.State.Keybinds {
+    using ColossalFramework;
```
Actually it is StyleCop warning SA1200. If we accept making StyleCop happy on this warning, then more usings will move inside, rather than staying outside.
![image](https://user-images.githubusercontent.com/288724/60534570-6a465380-9d02-11e9-9fd8-5bf86500a3a8.png)

aubergine10 replied:
It's certainly more semantic to have them inside the namespace. Although personally I find it easier to read them outside. It's not really an issue for now, I guess once harmony patch is done we'll be in better position to to a mass code cleanup so can address this when that happens.

kvakvs commented:
@aubergine10 
> I noticed in code some stuff about row striping, but didn't see any on the UI? (not important, just mentioning it in case the rows were supposed to be striped)

The row striping code is still there. But in the old dialog the default background was grey-ish, and row striping code just cleared texture on some lines. In the new dialog i've no idea what the background sprite name is, so they are all black (i.e. not striped). Have to go into ModTool or UIHelper source and find the background sprite name which is used in other dialogs.

aubergine10 commented:
LGTM!

@kvakvs Think we can merge this now - but there's some inline tasks (eg. in OP) - do you want to carry those over to another issue (eg. #402 )?

aubergine10 commented:
Resolved merge conflict that was caused by #400

kvakvs commented:
@aubergine10 You've already moved that +click task into #402 so that kind of removes it from here
